src/daemon/led_hw.c

```diff
@@ -75,26 +131,100 @@ void cmd_hwanim(usbdevice* kb, usbmode* dummy1, int dummy2, int keyindex, const
+            return;
+    }
+
+    if (IS_K63_WL(kb)) {
+        // K63 WL    Format: animation mode:speed:direction
+        uchar anim = HWANIM_NONE;
+        uchar speed = HWANIM_SPEED_NONE;
+        uchar direction = HWANIM_DIRECTION_NONE;
+        uchar opacity = 100; //100%
+        sscanf(code, "%hhx:%hhu:%hhu", &anim, &speed, &direction);
+
+        switch (anim) {
+        case HWANIM_COLOURSHIFT:
+        case HWANIM_COLOURPULSE:
+        case HWANIM_RAINBOW:
+        case HWANIM_VISOR:
+        case HWANIM_RAIN:
+        case HWANIM_TYPELIGHTNING_KEY:
+        case HWANIM_TYPELIGHTNING_RIPPLE:
+            switch (speed) {
+            case HWANIM_SPEED_LOW:
+            case HWANIM_SPEED_MEDIUM:
+            case HWANIM_SPEED_HIGH:
+                break;
+            default:
+                return;
+            }
+            //direction not supported
+            direction = HWANIM_DIRECTION_NONE;
+            break;
+        case HWANIM_STATIC:
+        case HWANIM_NONE:
+            //direction and speed not supported
+            direction = HWANIM_DIRECTION_NONE;
+            speed = HWANIM_SPEED_NONE;
+            break;
+        case HWANIM_WAVE:
+            switch (direction) {
+            case HWANIM_DIRECTION_LEFT:
+            case HWANIM_DIRECTION_RIGHT:
+                break;
+            default:
+                return;
+            }
+            break;
+        default:
             return;
-    } else if (IS_K63_WL(kb)) {
-        uchar data_pkt[MSG_SIZE] = {
+        }
+
+        uchar data_pkt[MSG_SIZE] ={
             CMD_SET, 0xaa, 0
         };
 
-        data_pkt[4]  = 0;
-        data_pkt[5]  = anim;
-        data_pkt[8]  = 100;
-        data_pkt[9]  = r[0];
-        data_pkt[10] = g[0];
-        data_pkt[11] = b[0];
-        // Second color for effects like alternating color shift
-        data_pkt[12] = r[0];
-        data_pkt[13] = g[0];
-        data_pkt[14] = b[0];
-        data_pkt[15] = 0;
+        //iCUE always changes the animation in 2 steps, so let's do the same
+
+        data_pkt[3]  = 0;
+        data_pkt[4]  = HWANIM_NONE;
+        data_pkt[5]  = HWANIM_SPEED_NONE;
+        data_pkt[6]  = HWANIM_DIRECTION_NONE;
+        data_pkt[7]  = 100;
+
+        if (!usbsend(kb, data_pkt, 1))
+            return -1;
+
+        data_pkt[3]  = 0;
+        data_pkt[4]  = anim;
+        data_pkt[5]  = speed;
+        data_pkt[6]  = direction;
+        data_pkt[7]  = 100;
+
+        if (!usbsend(kb, data_pkt, 1))
+            return -1;
+
+        //Optional:
+
+        //Turn on the light to ensure the animation is visible
+        //TODO: support it via the common way (how?)
+
+        /*
+            What's interesting, the hardware button for backlight intensity
+            on K63WL just sends the keypress code to Windows, then the iCUE
+            responds with change light command below.
+        */
+
+        uchar level = 1; //1,2,3 where l3 can be globally disabled
```

Comment:
What exactly does this do?

Reply:
hmm, good point, should be light_intensity_level